# Intro
I wanted to thrive in yahoo's fantasy basketball's 9-cat game.  
In the 9-cat game, I'm a "drafter" drafting basketball "players".  
As a drafter, we draft 13 players in total in a snake-draft with the other 11 drafters (12 drafters in total).   
We compete against other drafter's selected team.  
The 9 categories to compete against an opponent drafter in a 9-cat game are:  
- FG%
- FT%
- 3PTM
- Points
- Assists
- Rebounds
- Steals
- Blocks
- Turnovers
Win at least 5 categories accumlated in a whole week against an opponent drafter.   

By looking at players stats everyday costs me too much time.  
I want to design a project that can assist with my decision making in drafting players.  

This Project aims at preprocessing raw data scraped from https://www.basketball-reference.com/  
Main steps in this project: 
- Data Scraping and Preprocessing:
    - Includes EDA, filtering and joining tables using pandas.   
- Create a Team Roster Auto-Builder, based on:  
    - Weekly Player Performance Prediction.  
    - Dynamic Programming for team roster selection.  


Notes: 
Fantasy basketball drafting is actually such a complex contrained optimization problem that I would like to dig deep in another project (Multiple Knapsack Assignment Problem).  
Right now I just want to "assist" my decision making rather than creating a whole automated roster building algorithm.  



# 1. Data Scraping and Preprocessing

In [2]:
# Use basketball_reference_web_scraper package.  https://jaebradley.github.io/basketball_reference_web_scraper/  
from basketball_reference_web_scraper import client
from basketball_reference_web_scraper.data import OutputType
import pandas as pd

In [3]:
# players_season_totals
df_pst2018=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2018))
df_pst2019=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2019))
df_pst2021=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2021))
df_pst2022=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2022))
df_pst2023=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2023))
df_pst2024=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2024))

In [4]:
# view the first 5 rows of the 2018 data  
df_pst2018.head()

slug           name                  positions  age  \
0  abrinal01   Álex Abrines  [Position.SHOOTING_GUARD]   24   
1    acyqu01     Quincy Acy   [Position.POWER_FORWARD]   27   
2  adamsst01   Steven Adams          [Position.CENTER]   24   
3  adebaba01    Bam Adebayo          [Position.CENTER]   20   
4  afflaar01  Arron Afflalo  [Position.SHOOTING_GUARD]   32   

                         team  games_played  games_started  minutes_played  \
0  Team.OKLAHOMA_CITY_THUNDER            75              8            1134   
1          Team.BROOKLYN_NETS            70              8            1359   
2  Team.OKLAHOMA_CITY_THUNDER            76             76            2487   
3             Team.MIAMI_HEAT            69             19            1368   
4          Team.ORLANDO_MAGIC            53              3             682   

   made_field_goals  attempted_field_goals  ...  made_free_throws  \
0               115                    291  ...                39   
1               130                    365  ...                49   
2               448                    712  ...               160   
3               174                    340  ...               129   
4                65                    162  ...                22   

   attempted_free_throws  offensive_rebounds  defensive_rebounds  assists  \
0                     46                  26                  88       28   
1                     60                  40                 217       57   
2                    286                 384                 301       88   
3                    179                 118                 263      101   
4                     26                   4                  62       30   

   steals  blocks  turnovers  personal_fouls  points  
0      38       8         25             124     353  
1      33      29         60             149     411  
2      92      78        128             215    1056  
3      32      41         66             138     477  
4       4       9         21              56     179  

[5 rows x 22 columns]

Data cleaning and aggregation.  
The raw data only have columns made_field_goals, attempted_field_goals.  We want the aggregated result of made_field_goals/attempted_field_goals to get the Field Goal Percentage (FGperc)  
We also want to convert the total accumulated data into daily average data.  

In [5]:
df2018 = df_pst2018
df2018 = df2018.drop(columns=['positions', 'team'])
df2018 = df2018.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2018['FGperc'] = df2018['made_field_goals'] / df2018['attempted_field_goals']
df2018['FTperc'] = df2018['made_free_throws'] / df2018['attempted_free_throws']
df2018['rebs'] = df2018['offensive_rebounds'] + df2018['defensive_rebounds']
df2018['ppg'] = df2018['points'] / df2018['games_played']
df2018['rpg'] = df2018['rebs'] / df2018['games_played']
df2018['apg'] = df2018['assists'] / df2018['games_played']
df2018['spg'] = df2018['steals'] / df2018['games_played']
df2018['bpg'] = df2018['blocks'] / df2018['games_played']
df2018['topg'] = df2018['turnovers'] / df2018['games_played']
df2018['3pg'] = df2018['made_three_point_field_goals'] / df2018['games_played']

c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


Let's have a brief understanding of what value of each categories count as good or bad.  
We use the quantiles of the data to determine three levels: elite/average/poor.  
We do it on the years 2018~2024.  
Note that data in year 2024 represents the current season that is update everyday.  

In [6]:
print("\n2018 (within top 120)")
print(f'FG%: elite: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2018["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2018["topg"].quantile(0.5).round(decimals=3)}, poor: {df2018["topg"].quantile(0.75).round(decimals=3)}')

# accumulate (low total: blk, stl, turnover) -> total
# average (high total: points, ast, 3pg, reb) -> top 120


2018 (within top 120)
FG%: elite: 0.591,       avg: 0.56, poor: 0.524
FT%: elite: 0.904,       avg: 0.868, poor: 0.85
3PTM: elite: 2.318,       avg: 1.925, poor: 1.648
Points: elite: 19.427,       avg: 15.937, poor: 13.874
Assists: elite: 5.229,       avg: 3.993, poor: 3.191
Rebounds: elite: 7.665,       avg: 6.305, poor: 5.356
Steals: elite: 1.462,       avg: 1.149, poor: 1.028
Blocks: elite: 1.098,       avg: 0.828, poor: 0.652
Turnovers: elite: 0.485,       avg: 0.94, poor: 1.45


In [7]:
df2019 = df_pst2019
df2019 = df2019.drop(columns=['positions', 'team'])
df2019 = df2019.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2019['FGperc'] = df2019['made_field_goals'] / df2019['attempted_field_goals']
df2019['FTperc'] = df2019['made_free_throws'] / df2019['attempted_free_throws']
df2019['rebs'] = df2019['offensive_rebounds'] + df2019['defensive_rebounds']
df2019['ppg'] = df2019['points'] / df2019['games_played']
df2019['rpg'] = df2019['rebs'] / df2019['games_played']
df2019['apg'] = df2019['assists'] / df2019['games_played']
df2019['spg'] = df2019['steals'] / df2019['games_played']
df2019['bpg'] = df2019['blocks'] / df2019['games_played']
df2019['topg'] = df2019['turnovers'] / df2019['games_played']
df2019['3pg'] = df2019['made_three_point_field_goals'] / df2019['games_played']

print("\n2019 (within top 120)")
print(f'FG%: elite: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2019["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2019["topg"].quantile(0.5).round(decimals=3)}, poor: {df2019["topg"].quantile(0.75).round(decimals=3)}')


2019 (within top 120)
FG%: elite: 0.595,       avg: 0.548, poor: 0.518
FT%: elite: 0.889,       avg: 0.864, poor: 0.841
3PTM: elite: 2.313,       avg: 1.911, poor: 1.611
Points: elite: 20.683,       avg: 16.673, poor: 14.186
Assists: elite: 5.523,       avg: 4.029, poor: 3.279
Rebounds: elite: 8.272,       avg: 6.688, poor: 5.45
Steals: elite: 1.404,       avg: 1.132, poor: 1.0
Blocks: elite: 1.11,       avg: 0.823, poor: 0.666
Turnovers: elite: 0.529,       avg: 0.856, poor: 1.452


c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


In [8]:
df2021 = df_pst2021
df2021 = df2021.drop(columns=['positions', 'team'])
df2021 = df2021.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2021['FGperc'] = df2021['made_field_goals'] / df2021['attempted_field_goals']
df2021['FTperc'] = df2021['made_free_throws'] / df2021['attempted_free_throws']
df2021['rebs'] = df2021['offensive_rebounds'] + df2021['defensive_rebounds']
df2021['ppg'] = df2021['points'] / df2021['games_played']
df2021['rpg'] = df2021['rebs'] / df2021['games_played']
df2021['apg'] = df2021['assists'] / df2021['games_played']
df2021['spg'] = df2021['steals'] / df2021['games_played']
df2021['bpg'] = df2021['blocks'] / df2021['games_played']
df2021['topg'] = df2021['turnovers'] / df2021['games_played']
df2021['3pg'] = df2021['made_three_point_field_goals'] / df2021['games_played']

print("\n2021 (within top 120)")
print(f'FG%: elite: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2021["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2021["topg"].quantile(0.5).round(decimals=3)}, poor: {df2021["topg"].quantile(0.75).round(decimals=3)}')


2021 (within top 120)
FG%: elite: 0.618,       avg: 0.556, poor: 0.519
FT%: elite: 0.933,       avg: 0.891, poor: 0.868
3PTM: elite: 2.736,       avg: 2.216, poor: 1.931
Points: elite: 21.933,       avg: 17.595, poor: 14.91
Assists: elite: 5.82,       avg: 4.442, poor: 3.509
Rebounds: elite: 7.932,       avg: 6.659, poor: 5.727
Steals: elite: 1.282,       avg: 1.115, poor: 1.0
Blocks: elite: 1.142,       avg: 0.932, poor: 0.749
Turnovers: elite: 0.5,       avg: 0.89, poor: 1.423


c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


In [9]:
df2022 = df_pst2022
df2022 = df2022.drop(columns=['positions', 'team'])
df2022 = df2022.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2022['FGperc'] = df2022['made_field_goals'] / df2022['attempted_field_goals']
df2022['FTperc'] = df2022['made_free_throws'] / df2022['attempted_free_throws']
df2022['rebs'] = df2022['offensive_rebounds'] + df2022['defensive_rebounds']
df2022['ppg'] = df2022['points'] / df2022['games_played']
df2022['rpg'] = df2022['rebs'] / df2022['games_played']
df2022['apg'] = df2022['assists'] / df2022['games_played']
df2022['spg'] = df2022['steals'] / df2022['games_played']
df2022['bpg'] = df2022['blocks'] / df2022['games_played']
df2022['topg'] = df2022['turnovers'] / df2022['games_played']
df2022['3pg'] = df2022['made_three_point_field_goals'] / df2022['games_played']

print("\n2022 (within top 120)")
print(f'FG%: elite: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2022["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2022["topg"].quantile(0.5).round(decimals=3)}, poor: {df2022["topg"].quantile(0.75).round(decimals=3)}')


2022 (within top 120)
FG%: elite: 0.644,       avg: 0.571, poor: 0.533
FT%: elite: 0.976,       avg: 0.888, poor: 0.868
3PTM: elite: 2.695,       avg: 2.23, poor: 1.906
Points: elite: 21.183,       avg: 17.301, poor: 15.015
Assists: elite: 5.767,       avg: 4.439, poor: 3.561
Rebounds: elite: 8.318,       avg: 6.399, poor: 5.497
Steals: elite: 1.341,       avg: 1.152, poor: 1.017
Blocks: elite: 1.066,       avg: 0.829, poor: 0.692
Turnovers: elite: 0.469,       avg: 0.809, poor: 1.297


c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


In [10]:
df2023 = df_pst2023
df2023 = df2023.drop(columns=['positions', 'team'])
df2023 = df2023.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2023['FGperc'] = df2023['made_field_goals'] / df2023['attempted_field_goals']
df2023['FTperc'] = df2023['made_free_throws'] / df2023['attempted_free_throws']
df2023['rebs'] = df2023['offensive_rebounds'] + df2023['defensive_rebounds']
df2023['ppg'] = df2023['points'] / df2023['games_played']
df2023['rpg'] = df2023['rebs'] / df2023['games_played']
df2023['apg'] = df2023['assists'] / df2023['games_played']
df2023['spg'] = df2023['steals'] / df2023['games_played']
df2023['bpg'] = df2023['blocks'] / df2023['games_played']
df2023['topg'] = df2023['turnovers'] / df2023['games_played']
df2023['3pg'] = df2023['made_three_point_field_goals'] / df2023['games_played']

print("\n2023 (within top 120)")
print(f'FG%: elite: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2023["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2023["topg"].quantile(0.5).round(decimals=3)}, poor: {df2023["topg"].quantile(0.75).round(decimals=3)}')


2023 (within top 120)
FG%: elite: 0.638,       avg: 0.576, poor: 0.542
FT%: elite: 1.0,       avg: 0.894, poor: 0.867
3PTM: elite: 2.647,       avg: 2.109, poor: 1.91
Points: elite: 23.219,       avg: 19.475, poor: 15.352
Assists: elite: 6.102,       avg: 4.823, poor: 3.732
Rebounds: elite: 8.455,       avg: 6.426, poor: 5.462
Steals: elite: 1.318,       avg: 1.089, poor: 0.974
Blocks: elite: 1.033,       avg: 0.783, poor: 0.633
Turnovers: elite: 0.519,       avg: 0.88, poor: 1.5


c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


In [11]:
df2024 = df_pst2024
df2024 = df2024.drop(columns=['positions', 'team'])
df2024 = df2024.groupby(['slug', 'name', 'age'], as_index=False)['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals'].sum()
df2024['FGperc'] = df2024['made_field_goals'] / df2024['attempted_field_goals']
df2024['FTperc'] = df2024['made_free_throws'] / df2024['attempted_free_throws']
df2024['rebs'] = df2024['offensive_rebounds'] + df2024['defensive_rebounds']
df2024['ppg'] = df2024['points'] / df2024['games_played']
df2024['rpg'] = df2024['rebs'] / df2024['games_played']
df2024['apg'] = df2024['assists'] / df2024['games_played']
df2024['spg'] = df2024['steals'] / df2024['games_played']
df2024['bpg'] = df2024['blocks'] / df2024['games_played']
df2024['topg'] = df2024['turnovers'] / df2024['games_played']
df2024['3pg'] = df2024['made_three_point_field_goals'] / df2024['games_played']

print("\n2024 (within top 120)")
print(f'FG%: elite: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2024["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2024["topg"].quantile(0.5).round(decimals=3)}, poor: {df2024["topg"].quantile(0.75).round(decimals=3)}')

c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until



2024 (within top 120)
FG%: elite: 0.637,       avg: 0.569, poor: 0.528
FT%: elite: 1.0,       avg: 0.908, poor: 0.875
3PTM: elite: 2.737,       avg: 2.182, poor: 1.797
Points: elite: 22.627,       avg: 18.0, poor: 14.662
Assists: elite: 5.526,       avg: 4.483, poor: 3.641
Rebounds: elite: 8.425,       avg: 6.377, poor: 5.5
Steals: elite: 1.37,       avg: 1.105, poor: 1.0
Blocks: elite: 1.175,       avg: 0.828, poor: 0.684
Turnovers: elite: 0.394,       avg: 0.842, poor: 1.433


Now we got a brief understanding of what the data looks like, let's start some analytics.  

# 2. Create a Team Roster Auto-Builder

We divide the process it into 2 steps:  
1. Weekly Player Performance Prediction: Average.  
2. Dynamic Programming for team roster selection.  

# 2-1. Weekly Player Performance Prediction: Average

We will first use the average values to determine the "daily" average.  
Then we will muliply by the total games played that week for the player to get the weekly projection (average).  

Get Average 9-cat stat for each player for recent seasons.  
We want the data from years 2021~2024.  
If the player's a rookie, use only the current season's data (year=2024).  

Lets first ask: which records are in df2022 and not in df2021?

In [12]:
df2022[~df2022['name'].isin(df2021['name'])].sort_values(by='points', ascending=False).head(20)


slug              name  age  made_field_goals  \
558  wagnefr01      Franz Wagner   20               456   
205  greenja05       Jalen Green   19               406   
31   barnesc01    Scottie Barnes   20               459   
123  cunnica01   Cade Cunningham   20               429   
383  mobleev01       Evan Mobley   20               422   
381  mitchda01   Davion Mitchell   23               352   
292  joneshe01     Herbert Jones   23               274   
150  duartch01      Chris Duarte   24               268   
259  hylanbo01      Bones Hyland   21               230   
496  sengual01    Alperen Şengün   19               249   
142  dosunay01       Ayo Dosunmu   22               278   
189  giddejo01       Josh Giddey   19               281   
536  thompkl01     Klay Thompson   31               246   
317  kuminjo01  Jonathan Kuminga   19               236   
307  kispeco01     Corey Kispert   22               234   
346   manntr01          Tre Mann   20               231   
102  chrisjo01  Josh Christopher   20               224   
533  thomaca02        Cam Thomas   20               218   
521  suggsja01       Jalen Suggs   20               198   
589  willizi02   Ziaire Williams   20               191   

     attempted_field_goals  made_free_throws  attempted_free_throws  \
558                    975               189                    219   
205                    954               188                    236   
31                     932               158                    215   
123                   1031               142                    168   
383                    830               167                    252   
381                    843                56                     85   
292                    576               137                    163   
150                    621                90                    112   
259                    571               107                    125   
496                    525               165                    232   
142                    535                53                     78   
189                    670                56                     79   
536                    573                46                     51   
317                    460               130                    190   
307                    514                54                     62   
346                    588                65                     82   
102                    500                83                    113   
533                    503                87                    105   
521                    548               126                    163   
589                    424                43                     55   

     offensive_rebounds  defensive_rebounds  games_played  ...    FGperc  \
558                  84                 272            79  ...  0.467692   
205                  33                 193            67  ...  0.425577   
31                  195                 362            74  ...  0.492489   
123                  58                 296            64  ...  0.416101   
383                 143                 427            69  ...  0.508434   
381                  33                 133            75  ...  0.417556   
292                  98                 198            78  ...  0.475694   
150                  41                 185            55  ...  0.431562   
259                  14                 174            69  ...  0.402802   
496                 138                 255            72  ...  0.474286   
142                  27                 187            77  ...  0.519626   
189                  98                 323            54  ...  0.419403   
536                  15                 110            32  ...  0.429319   
317                  54                 180            70  ...  0.513043   
307                  53                 156            77  ...  0.455253   
346                  25                 146            60  ...

155 players in 2022 didn't appear in 2021: wagnefr01 Franz Wagner, thompkl01 Klay Thompson, ...  
They happen to be rookies in 2022 or players that were injured in the previous year (and now recovered).  

In [13]:
df2021[~df2021['name'].isin(df2022['name'])].sort_values(by='points', ascending=False).head(20)

slug                     name  age  made_field_goals  \
527  willizi01          Zion Williamson   20               634   
292  leonaka01            Kawhi Leonard   29               465   
358  murraja01             Jamal Murray   23               378   
456  simmobe01              Ben Simmons   24               325   
508   walljo01                John Wall   30               293   
371   nunnke01            Kendrick Nunn   25               319   
23   bacondw01             Dwayne Bacon   25               292   
532  wisemja01            James Wiseman   19               191   
446  saricda01              Dario Šarić   26               153   
470  sumneed01            Edmond Sumner   25               148   
473  teaguje01              Jeff Teague   32               123   
422  redicjj01              J.J. Redick   36               106   
34   baynear01              Aron Baynes   34               134   
93   cartemi01  Michael Carter-Williams   29               105   
419  randlch01           Chasson Randle   27                92   
170  gasolma01               Marc Gasol   36                88   
449  scottmi01               Mike Scott   32                76   
393  pattepa01        Patrick Patterson   31                72   
376  okafoja01            Jahlil Okafor   25                63   
249  jamesju01             Justin James   24                52   

     attempted_field_goals  made_free_throws  attempted_free_throws  \
527                   1037               369                    529   
292                    908               261                    295   
358                    792               133                    153   
456                    583               176                    287   
508                    726               158                    211   
371                    658                56                     60   
23                     726               145                    176   
532                    368                54                     86   
446                    342                84                     99   
470                    282                68                     83   
473                    284                89                    105   
422                    267                49                     52   
34                     304                29                     41   
93                     270                49                     80   
419                    237                38                     48   
170                    194                36                     50   
449                    211                 8                     12   
393                    165                13                     17   
376                    102                17                     24   
249                    111                21                     36   

     offensive_rebounds  defensive_rebounds  games_played  ...    FGperc  \
527                 167                 274            61  ...  0.611379   
292                  55                 282            52  ...  0.512115   
358                  37                 157            48  ...  0.477273   
456                  93                 324            58  ...  0.557461   
508                  17                 112            40  ...  0.403581   
371                  21                 158            56  ...  0.484802   
23                   31                 193            72  ...  0.402204   
532                  55                 171            39  ...  0.519022   
446                  46                 144            50  ...  0.447368   
470                  18                  76            53  ...  0.524823   
473                  15                  73            55  ...  0.433099   
422                   4                  60            44  ...  0.397004   
34                   87                 186            53  ...  0.440789   
93                   39                 100            31  ...  0.3

90 players in 2021 didn't appear in 2022: willizi01	Zion Williamson, leonaka01	Kawhi Leonard, ...  
If we want the data from year 2021~2024, find the Set Union of those 4 years's player list where we:  
    concat -> groupby -> average (weighted).   

We want a weighted average of year 2021~2024 since year 2024 should weigh more than older years like 2021.  

In [14]:
# Weighted Sum of average stats: (We can add a 'weights' attribute to each years df. )
    # (0.45*2023stat + 0.35*2022stat + 0.2*2021stat) / (0.45*2023gamesplayed + 0.35*2022gamesplayed + 0.2*2021gamesplayed)
# 'slug', 'name', (9-cat weighted sum)... 

# set weights for each year
wts = {"2024": 0.5,
       "2023": 0.25, 
       "2022": 0.15, 
       "2021": 0.1}

df2024['wts'] = wts['2024']
df2023['wts'] = wts['2023']
df2022['wts'] = wts['2022']
df2021['wts'] = wts['2021']

def weighted_average(group, values, games_played, weight):
    tmp=[]
    gp = group[games_played]
    wt = group[weight]
    for value in values:
        val = group[value]        
        tmp.append((val * wt).sum() / (gp*wt).sum())
    return pd.Series(tmp, index=values)

values=['made_three_point_field_goals', 'points', 'assists', 'rebs', 'steals', 'blocks', 'turnovers']

# Although outer join is a little faster, it will miss the rows with the "same exact values". Concatenation wouldn't miss.
df_pred_avg_21_24 = pd.concat([df2021, df2022, df2023, df2024]).fillna(0)[['slug', 'name', *values, 'games_played', 'wts']] \
    .groupby(['slug', 'name'], as_index=False) \
    .apply(weighted_average, values, 'games_played', 'wts')


Let's have a look at the results of famous players.  

In [15]:
df_pred_avg_21_24[df_pred_avg_21_24['name'].isin(['Klay Thompson', 'Stephen Curry', 'Kevin Durant', 'LeBron James', 'Nikola Jokić', \
                            'Giannis Antetokounmpo', 'Tyrese Haliburton', 'De\'Aaron Fox', 'Victor Wembanyama'])]

slug                   name  made_three_point_field_goals  \
15   antetgi01  Giannis Antetokounmpo                      0.799517   
172  curryst01          Stephen Curry                      4.889460   
212  duranke01           Kevin Durant                      2.123077   
244    foxde01           De'Aaron Fox                      1.805838   
295  halibty01      Tyrese Haliburton                      2.744600   
380  jamesle01           LeBron James                      2.368349   
399  jokicni01           Nikola Jokić                      1.108672   
754  thompkl01          Klay Thompson                      3.767036   
808  wembavi01      Victor Wembanyama                      1.444444   

        points   assists       rebs    steals    blocks  turnovers  
15   30.102657  5.571256  11.396135  1.062802  1.128019   3.733092  
172  28.778920  5.832905   5.526992  1.026992  0.289203   3.314910  
212  29.576923  5.627692   6.830769  0.775385  1.232308   3.435385  
244  25.498731  6.242386   4.110406  1.272843  0.373096   2.559645  
295  19.274460  9.297332   3.664549  1.500635  0.556544   2.396442  
380  27.512960  6.673943   8.013643  1.139154  0.732606   3.470668  
399  26.397366  9.110867  12.351262  1.308452  0.746432   3.405049  
754  19.795563  2.391442   4.017433  0.591125  0.484945   1.781300  
808  19.277778  2.555556   9.666667  1.333333  2.722222   3.500000

However, these are "daily" stats rather than "weekly".  

**Weekly Projection for All Players:**  
Because we compete against a new opponent every week, we need a player's weekly performance prediction.  
Therefore, we need to take game schedule in to consideration.  
- Weekly(Monday~Sunday) Projection = Daily Prediction(use 'Average' for now) * games played   

We will create a function getWeekProjection_2324() at the end of this subsection.  

First get the 2023-2024 season schedule from bball ref.   

In [16]:
from basketball_reference_web_scraper import client
df_schedule2024 = pd.DataFrame.from_dict(client.season_schedule(season_end_year=2024))
df_schedule2024.head()

start_time                away_team  \
0 2023-10-24 23:30:00+00:00  Team.LOS_ANGELES_LAKERS   
1 2023-10-25 02:00:00+00:00        Team.PHOENIX_SUNS   
2 2023-10-25 23:00:00+00:00     Team.HOUSTON_ROCKETS   
3 2023-10-25 23:00:00+00:00      Team.BOSTON_CELTICS   
4 2023-10-25 23:00:00+00:00  Team.WASHINGTON_WIZARDS   

                    home_team  away_team_score  home_team_score  
0         Team.DENVER_NUGGETS            107.0            119.0  
1  Team.GOLDEN_STATE_WARRIORS            108.0            104.0  
2          Team.ORLANDO_MAGIC             86.0            116.0  
3        Team.NEW_YORK_KNICKS            108.0            104.0  
4         Team.INDIANA_PACERS            120.0            143.0

Get Player List of 2024 and join with the current team name.  
Traded Players have multiple rows in df_pst2024.  
For example, James Harden was traded at the beginning of the season so he has two records: one with the 76ers, and one with the Clippers.  

We only need to know the newest team of that player.  

In [17]:
# df_pst2024 has 489, but only 483 unique names
print(df_pst2024.shape)
print(len(df_pst2024['name'].unique()))

# find the duplicate names ==> aka players that've been traded this season
df_tmp = df_pst2024[['name']].groupby(['name'], as_index=False).size()
traded_playerlist = list(df_tmp[df_tmp['size']>1]['name'])

df_pst2024[df_pst2024['name'].isin(traded_playerlist)]


(497, 22)
491


slug              name                 positions  age  \
32   batumni01     Nicolas Batum  [Position.POWER_FORWARD]   35   
33   batumni01     Nicolas Batum  [Position.POWER_FORWARD]   35   
94   covinro01  Robert Covington  [Position.POWER_FORWARD]   33   
95   covinro01  Robert Covington  [Position.SMALL_FORWARD]   33   
276  martike04         KJ Martin  [Position.SMALL_FORWARD]   23   
277  martike04         KJ Martin  [Position.SMALL_FORWARD]   23   
343  petrufi01    Filip Petrušev         [Position.CENTER]   23   
344  petrufi01    Filip Petrušev         [Position.CENTER]   23   
428  theisda01      Daniel Theis         [Position.CENTER]   31   
429  theisda01      Daniel Theis         [Position.CENTER]   31   
441  tuckepj01       P.J. Tucker  [Position.POWER_FORWARD]   38   
442  tuckepj01       P.J. Tucker  [Position.POWER_FORWARD]   38   

                          team  games_played  games_started  minutes_played  \
32   Team.LOS_ANGELES_CLIPPERS             3              0              54   
33     Team.PHILADELPHIA_76ERS            10              7             254   
94   Team.LOS_ANGELES_CLIPPERS             3              3              69   
95     Team.PHILADELPHIA_76ERS            16              3             287   
276  Team.LOS_ANGELES_CLIPPERS             2              0              31   
277    Team.PHILADELPHIA_76ERS             9              0              54   
343    Team.PHILADELPHIA_76ERS             1              0               3   
344      Team.SACRAMENTO_KINGS             2              0               7   
428        Team.INDIANA_PACERS             1              0               8   
429  Team.LOS_ANGELES_CLIPPERS             9              0             155   
441    Team.PHILADELPHIA_76ERS             3              3              66   
442  Team.LOS_ANGELES_CLIPPERS            12              0             173   

     made_field_goals  attempted_field_goals  ...  made_free_throws  \
32                  3                      8  ...                 0   
33                 24                     40  ...                 3   
94                  3                      9  ...                 1   
95                 31                     68  ...                17   
276                 4                     10  ...                 1   
277                 4                     12  ...                 2   
343                 0                      0  ...                 0   
344                 1                      2  ...                 1   
428                 1                      4  ...                 0   
429                30                     62  ...                 5   
441                 2                      5  ...                 0   
442                 4                     14  ...                 3   

     attempted_free_throws  offensive_rebounds  defensive_rebounds  assists  \
32                       0                   0                   7        5   
33                       4                  12                  29       16   
94                       2                   2                   6        7   
95                      19                  24                  37       11   
276                      2                   1                   2        1   
277                      4                   2                   7        3   
343                      0                   0                   1        0   
344                      2                   0                   0        0   
428                      0                   0                   0        0   
429                     10                  17                  21       12   
441                      0                   1                  13        0   
442                      3                  15                  14        8   

     steals  blocks  turnovers  personal_fouls  points  
32        3       4          1               5       8  
33        6       7          6              

Conclusion from above cell: rows that appear later (larger index) are the newer teams of that traded player.  

In [18]:

playerlist2024 = df_pst2024[['slug', 'name', 'team']]

playerlist2024['idx'] = range(0,len(playerlist2024))

# only keep the row with the larger index if rows having same names and slugs
playerlist2024 = playerlist2024.loc[playerlist2024.groupby(['slug', 'name'])['idx'].idxmax()].drop(columns=['idx'])
playerlist2024


c:\Users\joe_c\miniconda3\envs\basketball\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


slug                      name                         team
0    achiupr01          Precious Achiuwa         Team.TORONTO_RAPTORS
1    adebaba01               Bam Adebayo              Team.MIAMI_HEAT
2    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ
3    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES
4    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES
..         ...                       ...                          ...
492  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS
493  youngtr01                Trae Young           Team.ATLANTA_HAWKS
494  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ
495  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS
496  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS

[491 rows x 3 columns]

Now let's add the team information to the player projection dataframe by inner-joining these two dataframes:
1. playerlist2024 (team information)
2. df_pred_avg_21_24 (player projection)

In [19]:
df_weeklyproj_avg_24 = playerlist2024.merge(df_pred_avg_21_24, on=['slug', 'name'], how='inner')
df_weeklyproj_avg_24


slug                      name                         team  \
0    achiupr01          Precious Achiuwa         Team.TORONTO_RAPTORS   
1    adebaba01               Bam Adebayo              Team.MIAMI_HEAT   
2    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ   
3    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES   
4    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES   
..         ...                       ...                          ...   
486  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS   
487  youngtr01                Trae Young           Team.ATLANTA_HAWKS   
488  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ   
489  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS   
490  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS   

     made_three_point_field_goals     points    assists      rebs    steals  \
0                        0.493386   8.148148   1.103175  5.734127  0.510582   
1                        0.022864  20.220217   3.708785  9.468111  1.227437   
2                        1.282828   7.040404   1.121212  2.313131  0.404040   
3                        1.175123   9.306056   1.306056  4.805237  0.561375   
4                        1.336788   8.001295   2.230570  2.284974  0.740933   
..                            ...        ...        ...       ...       ...   
486                      0.177365   6.506757   2.194257  3.935811  0.979730   
487                      2.548387  26.799778  10.037820  3.278087  1.125695   
488                      0.135977   4.949008   0.784703  4.694051  0.257790   
489                      0.018957   5.028436   1.109005  4.504739  0.291469   
490                      0.002123  10.599788   1.151805  8.980892  0.340764   

       blocks  turnovers  
0    0.514550   1.075397  
1    0.877256   2.535499  
2    0.393939   0.717172  
3    0.589198   0.831424  
4    0.463731   1.155440  
..        ...        ...  
486  0.268581   1.097973  
487  0.100111   4.016685  
488  0.396601   0.688385  
489  0.218009   0.684834  
490  1.194268   1.421444  

[491 rows x 10 columns]

Now we want to know how many games does each team play this week?  

Let's get the 2023-2024 NBA game schedule.  
We also want to add a 'week' column for later grouping.  

In [20]:
import datetime

# first assume every player on the team plays. Just get the weekly schedule of the player's team. 
# view schema: team, week (starting from 2023-10-24), gamesplayed (during that week) 
# can also add columns: homegames, awaygames

df_schedule2024.start_time = df_schedule2024['start_time'].dt.tz_convert('America/Los_Angeles')
df_schedule2024.start_time = df_schedule2024.start_time.dt.date

# Create 'week' column: 
# week1: 2023-10-24 ~ 10-29, week2: 10-30 ~ 11-5, week3: 11-6 ~ 11-12, ...
# R dplyr::mutate, equivalent to pandas conditionally with list comprehension, or apply lambda

seasonStartDate = df_schedule2024.start_time[0]
seasonStartDate_weekday = datetime.timedelta(seasonStartDate.weekday())
df_schedule2024['week'] = df_schedule2024['start_time'].apply(lambda x: (x - seasonStartDate + seasonStartDate_weekday).days//7 + 1)
df_schedule2024


start_time                    away_team                    home_team  \
0     2023-10-24      Team.LOS_ANGELES_LAKERS          Team.DENVER_NUGGETS   
1     2023-10-24            Team.PHOENIX_SUNS   Team.GOLDEN_STATE_WARRIORS   
2     2023-10-25         Team.HOUSTON_ROCKETS           Team.ORLANDO_MAGIC   
3     2023-10-25          Team.BOSTON_CELTICS         Team.NEW_YORK_KNICKS   
4     2023-10-25      Team.WASHINGTON_WIZARDS          Team.INDIANA_PACERS   
...          ...                          ...                          ...   
1221  2024-04-14            Team.PHOENIX_SUNS  Team.MINNESOTA_TIMBERWOLVES   
1222  2024-04-14      Team.LOS_ANGELES_LAKERS    Team.NEW_ORLEANS_PELICANS   
1223  2024-04-14        Team.DALLAS_MAVERICKS   Team.OKLAHOMA_CITY_THUNDER   
1224  2024-04-14  Team.PORTLAND_TRAIL_BLAZERS        Team.SACRAMENTO_KINGS   
1225  2024-04-14         Team.DETROIT_PISTONS       Team.SAN_ANTONIO_SPURS   

      away_team_score  home_team_score  week  
0               107.0            119.0     1  
1               108.0            104.0     1  
2                86.0            116.0     1  
3               108.0            104.0     1  
4               120.0            143.0     1  
...               ...              ...   ...  
1221              NaN              NaN    25  
1222              NaN              NaN    25  
1223              NaN              NaN    25  
1224              NaN              NaN    25  
1225              NaN              NaN    25  

[1226 rows x 6 columns]

Group by week and team to know how many games does each team play this week?   

In [21]:
# groupby week and team
df_gamesplayed_weekly = df_schedule2024.copy()
df_gamesplayed_weekly['away_team'] = df_gamesplayed_weekly['away_team'].apply(lambda x: x.value)
df_gamesplayed_weekly['home_team'] = df_gamesplayed_weekly['home_team'].apply(lambda x: x.value)
df_gamesplayed_weekly

df_tmp1 = df_gamesplayed_weekly[['week', 'away_team']].groupby(['week', 'away_team'], as_index=False).size()
df_tmp2 = df_gamesplayed_weekly[['week', 'home_team']].groupby(['week', 'home_team'], as_index=False).size()
df_tmp1.rename(columns={'away_team':"team"}, inplace=True)
df_tmp2.rename(columns={'home_team':"team"}, inplace=True)
df_gamesplayed_weekly = pd.concat([df_tmp1, df_tmp2]).groupby(['week', 'team'], as_index=False).sum()
# df_gamesplayed_weekly.sort_values(by=['week', 'size'], ascending = [True, False])
df_gamesplayed_weekly


week                team  size
0       1       ATLANTA HAWKS     3
1       1      BOSTON CELTICS     2
2       1       BROOKLYN NETS     2
3       1   CHARLOTTE HORNETS     2
4       1       CHICAGO BULLS     3
..    ...                 ...   ...
745    25    SACRAMENTO KINGS     4
746    25   SAN ANTONIO SPURS     4
747    25     TORONTO RAPTORS     4
748    25           UTAH JAZZ     4
749    25  WASHINGTON WIZARDS     3

[750 rows x 3 columns]

Example: De'Aaron Fox's week6 projection  

In [22]:
# De'Aaron Fox week6 projection
week = 6
team = playerlist2024[playerlist2024.name=='De\'Aaron Fox'].team.values[0].value
gamesPlayed_week = df_gamesplayed_weekly[(df_gamesplayed_weekly['week']==week) & (df_gamesplayed_weekly['team']==team)].size
df_weeklyproj_avg_24[df_weeklyproj_avg_24['name']=='De\'Aaron Fox'].iloc[:,3:] * gamesPlayed_week

made_three_point_field_goals     points    assists       rebs    steals  \
131                      5.417513  76.496193  18.727157  12.331218  3.818528   

       blocks  turnovers  
131  1.119289   7.678934

Create a function getWeekProjection_2324()  
The input is an nba week number, and outputs are the weekly projection for all the players in that specified week.  
We have a look at the outputs sorted by made_three_point_field_goals.   

In [29]:
# func(week) ==> weekly projection of every player
def current_week():
    seasonStartDate = df_schedule2024.start_time[0]
    seasonStartDate_weekday = datetime.timedelta(seasonStartDate.weekday())
    week = (datetime.date.today() - seasonStartDate + seasonStartDate_weekday).days//7 + 1
    return week

def getWeekProjection_2324(week=current_week()):
    df_weeklyproj_avg_24 = playerlist2024.merge(df_pred_avg_21_24, on=['slug', 'name'], how='inner')
    gamesPlayed_week = df_gamesplayed_weekly[(df_gamesplayed_weekly['week']==week) & (df_gamesplayed_weekly['team']==team)].size
    df_weeklyproj_avg_24.iloc[:,3:] = df_weeklyproj_avg_24.iloc[:,3:] * gamesPlayed_week
    return df_weeklyproj_avg_24

df_tmp = getWeekProjection_2324(current_week())
df_tmp.sort_values(by='made_three_point_field_goals', ascending=False).head(20)




slug                name                         team  \
98   curryst01       Stephen Curry   Team.GOLDEN_STATE_WARRIORS   
427  thompkl01       Klay Thompson   Team.GOLDEN_STATE_WARRIORS   
254  lillada01      Damian Lillard         Team.MILWAUKEE_BUCKS   
183  hieldbu01         Buddy Hield          Team.INDIANA_PACERS   
302  mitchdo01    Donovan Mitchell     Team.CLEVELAND_CAVALIERS   
441  vanvlfr01       Fred VanVleet         Team.HOUSTON_ROCKETS   
20    ballla01         LaMelo Ball       Team.CHARLOTTE_HORNETS   
419  tatumja01        Jayson Tatum          Team.BOSTON_CELTICS   
109  doncilu01         Luka Dončić        Team.DALLAS_MAVERICKS   
140  georgpa01         Paul George    Team.LOS_ANGELES_CLIPPERS   
400  simonan01     Anfernee Simons  Team.PORTLAND_TRAIL_BLAZERS   
204  irvinky01        Kyrie Irving        Team.DALLAS_MAVERICKS   
165  hardati02    Tim Hardaway Jr.        Team.DALLAS_MAVERICKS   
34   beaslma01       Malik Beasley         Team.MILWAUKEE_BUCKS   
281  mccolcj01         CJ McCollum    Team.NEW_ORLEANS_PELICANS   
348  portemi01  Michael Porter Jr.          Team.DENVER_NUGGETS   
44   bogdabo01   Bogdan Bogdanović           Team.ATLANTA_HAWKS   
23    banede01        Desmond Bane       Team.MEMPHIS_GRIZZLIES   
377  roziete01        Terry Rozier       Team.CHARLOTTE_HORNETS   
271  markkla01     Lauri Markkanen               Team.UTAH_JAZZ   

     made_three_point_field_goals     points    assists       rebs    steals  \
98                      14.668380  86.336761  17.498715  16.580977  3.080977   
427                     11.301109  59.386688   7.174326  12.052298  1.773376   
254                     11.096961  86.227207  21.798842  13.593343  2.670043   
183                     10.321244  47.368912   8.754404  13.203109  2.906736   
302                     10.189460  81.609787  14.728984  13.404015  4.539523   
441                      9.697761  57.138060  21.951493  12.600746  4.597015   
20                       9.515982  64.018265  23.237443  18.639269  4.378995   
419                      9.347162  84.651747  13.211790  25.106987  3.193231   
109                      9.326165  91.544803  25.100358  25.763441  3.738351   
140                      9.008942  71.087928  14.923994  18.675112  4.694486   
400                      9.000000  51.751212  10.400646   7.550889  1.725363   
204                      8.995349  78.813953  16.855814  14.116279  3.558140   
165                      8.950695  45.758534   5.472819  10.433628  1.748420   
34                       8.944323  38.620087   4.729258  10.873362  2.338428   
281                      8.853061  64.546939  16.351020  12.967347  3.175510   
348                      8.609865  52.246637   3.973094  20.008969  1.816143   
44                       8.584803  45.020353   8.409769  10.115332  3.244233   
23                       8.417062  57.430095  11.090047  13.212085  3.191943   
377                      8.405512  61.862205  14.759843  12.496063  3.594488   
271                      8.376471  63.156863   4.345098  22.435294  2.223529   

       blocks  turnovers  
98   0.867609   9.944730  
427  1.454834   5.343899  
254  0.768452   9.160637  
183  1.165803   5.235233  
302  1.020075   8.183187  
441  1.638060   6.156716  
20   1.009132  10.319635  
419  1.748908   8.868996  
109  1.559140  11.856631  
140  0.970194   9.281669  
400  0.504039   5.190630  
204  1.860465   6.227907  
165  0.379267   2.541087  
34   0.409389   2.813319  
281  1.502041   6.269388  
348  1.923767   3.497758  
44   0.928087   3.635007  
23   1.276066   5.690758  
377  0.933071   5.685039  
271  1.725490   4.317647

Notes: In the 2023-2024 season, week7 is the week of In-Season Tournament that won't been scheduled beforehand. 

# 2-2. Dynamic Programming for team roster selection.  

13-roster Team Selection Mechanism under a 12-team Snake Draft.   
Create the 13-roster Team from the Weekly Projection using similar techinques as when solving the 0/1 knapsack problem.  

Find the best 13 players to win a 9-cat game regardless of the position constraint and the snake draft constraint.  

1. Brute Force: 13 players out of 483 (mylist) or 548 (fantasypros draft list). Using recursive has an O(2^m) time complexity, and C(483, 13) = 1.062392279 E+25 is too large of a number.   

2. Dynamic Programming: 0/1 Knapsack problem  
    Given N **items** where each item has some weight and **profit** associated with it and also given a bag with **capacity** W, [i.e., the bag can hold at most W weight in it]. The goal is to put the items into the bag such that is has the **max sum of profits** possible.  
    
    N: total player list (**items**)  
    W: 13-player roster team that consists of 1PG, 1SG, 1G, 1SF, 1PF, 1F, 2C, 2Util, 3BN (**capacity**)  
    profit: 9-cat stats (**profits**)  
    goal: categorical wins (**max sum of profits**)  




Let's first solve 0/1 knapsack problem.  

Assume that the 13-player team doesn't have position constraints.  
aka. every player's weight is 1, where capacity W=13.  

The difference in our case is that: **"profit" is multidimensional** since they're 9-cat stats.  
==> we can do an "operation overload" for this multidimensional-categorical profit  
    usually: profit1=8, profit2=4, profit3=2;  
        profit1 = 8 > profit2 + profit3 = 6  
    now: profit1 = [0,2,2,2,2,2,2,2,0], profit2 = [100,0,0,0,0,0,0,1,1], profit3 = [0,100,0,0,0,0,1,0,0]   
        profit1 = [0,2,2,2,2,2,2,2,2] > profit2 + profit3 = [100,100,0,0,0,0,1,1,1]   
        profit1 wins the last 8 categories over profit2 + profit3's first 2 categories. 
        list-elementwise greaterthan.   

In [24]:
# define the new greaterthan
def profit_greaterthan(A, B): 
    # A and B are lists. 
    l = []
    for (a, b) in zip(A,B):
        if a > b:
            l.append(1)
        elif a==b:
            l.append(0)
        else:
            l.append(-1)
    if sum(l) > 0:
        return True
    else:
        return False

def profit_max(A_exclude, B_include): 
    # A, B, and idx are lists. 
    if profit_greaterthan(A_exclude, B_include):
        return A_exclude
    else:
        return B_include
    
def profit_addition(A, B):
    # A and B are lists. 
    return [sum(x) for x in zip(A, B)]


# Space optimized Approach (code is inspired by Prosun Kumar Sarkar)
def knapSack(W, wt, val, n): 
    # Making the dp array each with 9 category values initialized as zero.  
    dp = [[0]*9 for i in range(W+1)]

    # Taking first i elements 
    for i in range(1, n+1): 
        # Starting from back, 
        # so that we also have data of 
        # previous computation when taking i-1 items 
        for w in range(W, 0, -1): 
            if wt[i-1] <= w: 
                # Find the max value 
                dp[w] = profit_max(dp[w], profit_addition(dp[w-wt[i-1]], val[i-1])) 
      
    # Returning the maximum value of knapsack 
    return dp[W]
  
  
profit = [[100,0,0,0,0,0,0,1,1], [0,100,0,0,0,0,1,0,0], [0,2,2,2,2,2,2,2,0]]
weight = [1, 1, 1]
W = 2
n = len(profit)
print(knapSack(W, weight, profit, n))


profit = [1, 7, 4]
weight = [3, 8, 2]
W = 6
N = 3





[100, 2, 2, 2, 2, 2, 2, 3, 1]


However, the above doesn't tell us which items were selected.  
We need to output the selected indexes.  

In [25]:
# define the new greaterthan
def profit_greaterthan(A, B): 
    # A and B are lists. 
    l = []
    for (a, b) in zip(A,B):
        if a > b:
            l.append(1)
        elif a==b:
            l.append(0)
        else:
            l.append(-1)
    if sum(l) > 0:
        return True
    else:
        return False

def profit_max(A_exclude, B_include): 
    # A, B, and idx are lists. 
    if profit_greaterthan(A_exclude, B_include):
        return A_exclude, 0
    else:
        return B_include, 1
    
def profit_addition(A, B):
    # A and B are lists. 
    return [sum(x) for x in zip(A, B)]


# Space optimized Approach (code is inspired by Prosun Kumar Sarkar)
def knapSack(W, wt, val, n): 
    # Making the dp array each with 9 category values initialized as zero.  
    dp = dp = [[[0]*9 if j==0 else [] for j in range(2)] for i in range(W+1)]

    # Taking first i elements 
    for i in range(1, n+1): 
        # Starting from back, 
        # so that we also have data of 
        # previous computation when taking i-1 items 
        for w in range(W, 0, -1): 
            if wt[i-1] <= w: 
                # Find the max value 
                dp[w][0], left_or_right = profit_max(dp[w][0], profit_addition(dp[w-wt[i-1]][0], val[i-1])) 
                if left_or_right==1: #include
                    dp[w][1] = dp[w-wt[i-1]][1].copy()
                    dp[w][1].append(i-1) 
                    
    # Returning the maximum value of knapsack 
    return dp[W]
  
  
profit = [[0,2,2,2,2,2,2,2,0], [0,100,0,0,0,0,1,0,0], [100,0,0,0,0,0,0,1,1]]
weight = [1, 1, 1]
W = 2
n = len(profit)
print(knapSack(W, weight, profit, n))



[[100, 2, 2, 2, 2, 2, 2, 3, 1], [0, 2]]


Try on our nba data df_currentweekproj = getWeekProjection_2324(current_week()). 

In [26]:
df_currentweekproj = getWeekProjection_2324(current_week())
df_currentweekproj

slug                      name                         team  \
0    achiupr01          Precious Achiuwa         Team.TORONTO_RAPTORS   
1    adebaba01               Bam Adebayo              Team.MIAMI_HEAT   
2    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ   
3    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES   
4    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES   
..         ...                       ...                          ...   
486  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS   
487  youngtr01                Trae Young           Team.ATLANTA_HAWKS   
488  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ   
489  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS   
490  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS   

     made_three_point_field_goals     points    assists       rebs    steals  \
0                        1.480159  24.444444   3.309524  17.202381  1.531746   
1                        0.068592  60.660650  11.126354  28.404332  3.682310   
2                        3.848485  21.121212   3.363636   6.939394  1.212121   
3                        3.525368  27.918167   3.918167  14.415712  1.684124   
4                        4.010363  24.003886   6.691710   6.854922  2.222798   
..                            ...        ...        ...        ...       ...   
486                      0.532095  19.520270   6.582770  11.807432  2.939189   
487                      7.645161  80.399333  30.113459   9.834260  3.377086   
488                      0.407932  14.847025   2.354108  14.082153  0.773371   
489                      0.056872  15.085308   3.327014  13.514218  0.874408   
490                      0.006369  31.799363   3.455414  26.942675  1.022293   

       blocks  turnovers  
0    1.543651   3.226190  
1    2.631769   7.606498  
2    1.181818   2.151515  
3    1.767594   2.494272  
4    1.391192   3.466321  
..        ...        ...  
486  0.805743   3.293919  
487  0.300334  12.050056  
488  1.189802   2.065156  
489  0.654028   2.054502  
490  3.582803   4.264331  

[491 rows x 10 columns]

In [27]:
df_profit = df_currentweekproj.iloc[:,3:]
df_profit.turnovers = df_profit.turnovers*-1
profit = df_profit.values.tolist()
n = len(profit)
weight = [1]*n
W = 13

values, player_idx = knapSack(W, weight, profit, n)
print(values)
print(player_idx)

df_currentweekproj.iloc[player_idx]

[100.8055140095579, 945.9403343631552, 239.49045045120187, 251.43442059074127, 50.4992464203401, 23.93759502511668, -103.9033934226306]
[98, 100, 109, 204, 213, 222, 249, 254, 281, 302, 311, 403, 441]


slug              name                        team  \
98   curryst01     Stephen Curry  Team.GOLDEN_STATE_WARRIORS   
100  davisan02     Anthony Davis     Team.LOS_ANGELES_LAKERS   
109  doncilu01       Luka Dončić       Team.DALLAS_MAVERICKS   
204  irvinky01      Kyrie Irving       Team.DALLAS_MAVERICKS   
213  jamesle01      LeBron James     Team.LOS_ANGELES_LAKERS   
222  jokicni01      Nikola Jokić         Team.DENVER_NUGGETS   
249  leonaka01     Kawhi Leonard   Team.LOS_ANGELES_CLIPPERS   
254  lillada01    Damian Lillard        Team.MILWAUKEE_BUCKS   
281  mccolcj01       CJ McCollum   Team.NEW_ORLEANS_PELICANS   
302  mitchdo01  Donovan Mitchell    Team.CLEVELAND_CAVALIERS   
311  murrade01   Dejounte Murray          Team.ATLANTA_HAWKS   
403  smartma01      Marcus Smart      Team.MEMPHIS_GRIZZLIES   
441  vanvlfr01     Fred VanVleet        Team.HOUSTON_ROCKETS   

     made_three_point_field_goals     points    assists       rebs    steals  \
98                      14.668380  86.336761  17.498715  16.580977  3.080977   
100                      0.964286  72.223214   8.821429  34.616071  3.312500   
109                      9.326165  91.544803  25.100358  25.763441  3.738351   
204                      8.995349  78.813953  16.855814  14.116279  3.558140   
213                      7.105048  82.538881  20.021828  24.040928  3.417462   
222                      3.326015  79.192097  27.332602  37.053787  3.925357   
249                      6.129964  69.958484  11.956679  18.974729  4.505415   
254                     11.096961  86.227207  21.798842  13.593343  2.670043   
281                      8.853061  64.546939  16.351020  12.967347  3.175510   
302                     10.189460  81.609787  14.728984  13.404015  4.539523   
311                      5.074610  59.669265  19.433185  17.906459  5.064588   
403                      5.378453  36.140884  17.639503   9.816298  4.914365   
441                      9.697761  57.138060  21.951493  12.600746  4.597015   

       blocks  turnovers  
98   0.867609   9.944730  
100  6.776786   6.357143  
109  1.559140  11.856631  
204  1.860465   6.227907  
213  2.197817  10.412005  
222  2.239297  10.215148  
249  1.689531   5.090253  
254  0.768452   9.160637  
281  1.502041   6.269388  
302  1.020075   8.183187  
311  0.761693   6.761693  
403  1.056630   7.267956  
441  1.638060   6.156716

EXPERIMENT: "choosing the top 13 players" vs. "using knapsack algorithm".  

In [28]:
# EXPERIMENT
# try simple sorting. getting the top 13. 

df_profit = df_currentweekproj.iloc[:,3:]
df_profit.turnovers = df_profit.turnovers*-1
profit = df_profit.values.tolist()

def function_top13(P):
    top13 = []
    visited = []
    for i in range(13):
        max_p = [0]*9
        max_idx = -1
        for j, p in enumerate(P):
            if j in visited:
                continue
            if profit_greaterthan(p, max_p):
                max_p = p
                max_idx = j
        visited.append(max_idx)
        top13.append(max_idx)
    return top13

# without using knapsack algo (Top 13).
tmp1 = df_currentweekproj.iloc[function_top13(profit)]
profit_simple = [0]*7
for p in tmp1.iloc[:, 3:].values.tolist():
    profit_simple = profit_addition(profit_simple, p)

# with using knapsack algo.
tmp2 = df_currentweekproj.iloc[player_idx]
profit_knapsack = [0]*7
for p in tmp2.iloc[:, 3:].values.tolist():
    profit_knapsack = profit_addition(profit_knapsack, p)

profit_greaterthan(profit_simple, profit_knapsack)

False

We can see that the output is False, which means using the 0/1 knapsack algorithm has a greater categorical profit than just picking the top 13 players.  

Discussion:

1. Weekly Player Performance Prediction: Time Series Predictions.  
    - I anticipate to have better results than using weighted average when using Time Series predictions, such as OLS Regression, XGBoost, Arima, Vector Autoregression, RNN/LSTM/GRU...  

2. Position constraint: 1PG, 1SG, 1G, 1SF, 1PF, 1F, 2C, 2Util, 3BN  
    -  This becomes a much more complex constraint optimization problem such as Mulitple Knapsack Assignment Problem (MKAP)

3. Snake Draft constraint: 
    - Simulate the snake draft: for simplicity, round1 only drafts within 1~12 picks, round2: 12~23, round3: 24~35,...   
    - Build a live draft version.  

4. Consider Accidents: Player Injuries, Game Time Decisions, Suspensions ...